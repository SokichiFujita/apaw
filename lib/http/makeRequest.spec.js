const https = require('https');
const makeRequest = require('./makeRequest');

jest.mock('https');

describe('makeRequest', () => {
  const resListeners = {};

  const reqEndSpy = jest.fn();
  const reqWriteSpy = jest.fn();
  const resOnSpy = jest.fn((a, b) => {
    resListeners[a] = b;
  });

  const resMock = {
    headers: {},
    statusCode: 200,
    on: resOnSpy,
  };

  https.request = jest.fn((options, callback) => {
    callback(resMock);
    return {
      end: reqEndSpy,
      write: reqWriteSpy,
    };
  });

  afterEach(() => {
    reqEndSpy.mockClear();
    reqWriteSpy.mockClear();
    resOnSpy.mockClear();
  });

  it('should make an HTTP request and resolve a promise', (done) => {
    const response = makeRequest({
      authorization: 'authorization',
      headers: {},
      host: 'webservices.amazon.com',
      path: '/',
      payload: '',
    });

    // mock some HTTP request response data
    resListeners.data('some');
    resListeners.data('data');
    resListeners.end();

    expect(reqEndSpy).toHaveBeenCalledTimes(1);
    expect(reqWriteSpy).toHaveBeenCalledTimes(1);
    expect(resOnSpy).toHaveBeenCalledTimes(3);

    response.then((data) => expect(data).toMatchSnapshot());
    done();
  });

  it('should make an HTTP request and reject a promise', (done) => {
    const mockError = new Error();
    const response = makeRequest({
      authorization: 'authorization',
      headers: {},
      host: 'webservices.amazon.com',
      path: '/',
      payload: '',
    });

    // mock an HTTP request response error
    resListeners.error(mockError);

    expect(reqEndSpy).toHaveBeenCalledTimes(1);
    expect(reqWriteSpy).toHaveBeenCalledTimes(1);
    expect(resOnSpy).toHaveBeenCalledTimes(3);

    response.catch((e) => expect(e).toMatchSnapshot());
    done();
  });
});
