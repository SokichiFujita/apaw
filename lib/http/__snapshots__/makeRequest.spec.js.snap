// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeRequest should make an HTTP request and reject a promise 1`] = `[Error]`;

exports[`makeRequest should make an HTTP request and resolve a promise 1`] = `"somedata"`;
